Port to 1.21.6-rc1

shartte opened the pull request:
Known issues and TODOs:

- [x] GuiGraphics: Add system to submit custom picture in picture renderers and custom render state elements for custom GUI rendering code (#2305)
- [ ] Since GuiGraphics draw calls are now collected and dispatched later, any global state set on the render system no longer affects them. This includes render pipeline modifiers. We will likely need to implement a stack of pipeline modifiers that is local to GuiGraphics (which gets applied to the pipeline before the render state element is recorded)
- [x] Fix info bars not rendering anymore (https://github.com/neoforged/NeoForge/pull/2297/commits/da1ac6020891c584870222ee36464db61efca60c)
- [x] Fix an old bugfix for [MC-262974](https://bugs.mojang.com/browse/MC/issues/MC-262974) getting lost in the port (https://github.com/neoforged/NeoForge/pull/8)

shartte committed:
Port Add source entity to MobEffectEvent.Applicable (#2276) (9f6f1812)

shartte committed:
Port Allow blocks full control over step, fall and hit sounds (#2284) (36df3bdf)

shartte committed:
Port Refactor the Anvil Events (#2260) (2d36447c)

shartte committed:
Port to 1.21.6-pre1

Co-authored-by: XFactHD <xfacthd@gmx.de>
Co-authored-by: Matyrobbrt <matyrobbrt@gmail.com>
Co-authored-by: coehlrich <coehlrich@users.noreply.github.com>

shartte committed:
Always register milk during datagen.

shartte committed:
Update data files.

shartte committed:
Fix TagBuilder addOptional bug.

Fuzss commented:
Info bars do not render. Neither experience bar nor jump meter do. Not sure how to test the new locator bar. 
Tested in singleplayer. 
If this is the wrong place to report please let me know where the appropriate place would be.

![2025-06-03_14 53 06](https://github.com/user-attachments/assets/256f063a-5da6-4e56-9290-129831686709)
![2025-06-03_14 54 12](https://github.com/user-attachments/assets/80309b85-77db-4cb2-acad-649da17ebcc2)

XFactHD committed:
Move GuiLayerManager.Layer out of internal class

shartte commented:
> Info bars do not render. Neither experience bar nor jump meter do. Not sure how to test the new locator bar. Tested in singleplayer. If this is the wrong place to report please let me know where the appropriate place would be.

Thanks for letting us know, that is very much a porting bug.

XFactHD committed:
Fix info bars not rendering properly

shartte committed:
Port to 1.21.6-pre2

shartte committed:
Port to 1.21.6-pre3

XFactHD committed:
[1.21.6] Add support for custom GUI element render states and PiP renderers (#2305)

shartte committed:
Run Datagen

shartte committed:
Re-fix https://github.com/neoforged/NeoForge/pull/8

XFactHD committed:
Add missing AT

shartte committed:
Disable natural mob spawning and weather in the game test server to prevent random interference of slimes with tests (#2314)

XFactHD committed:
Add ValueInput/ValueOutput-based serialization to ItemStackHandler, FluidTank and EnergyStorage
The raw NBT serialization of EnergyStorage is changed to serialize to a CompoundTag instead of an IntTag

Matyrobbrt committed:
Small changes after patch review

Matyrobbrt committed:
Clean-up of some (porting) TODOs

Matyrobbrt committed:
Port to 1.21.6-pre4

Fuzss commented:
Patches for firing `RenderBlockScreenEffectEvent` appear to be missing. At least IDE can't find any calls for it.

Matyrobbrt committed:
Make all removal entries optional at runtime (#2322)

lukebemish committed:
Avoid patching LanServerPinger into the dedicated server, using a neo-supplied implementation instead (#2308)

Disable advertising the dedicated server to local LAN clients in production.

thiakil committed:
Refactor the living entity damage tests to detect prior damage and be much faster (#2312)

Co-authored-by: shartte <shartte@users.noreply.github.com>

XFactHD commented:
All patches for screen effects are present: https://github.com/neoforged/NeoForge/blob/a4de4e27aae59076b960479465e4294ad13e8337/patches/net/minecraft/client/renderer/ScreenEffectRenderer.java.patch

Fuzss commented:
You're right, thank you. Not sure why I couldn't see the patches.

Matyrobbrt committed:
Bump ModLauncher to 11.0.5 to catch early module loading errors

See https://github.com/McModLauncher/modlauncher/pull/134

Matyrobbrt committed:
Port to 1.21.6-rc1

Matyrobbrt committed:
Move attachment serialisers from raw NBT to ValueI/O (#2316)

lukebemish committed:
[1.21.6] [Breaking] Fix lambda names in production and enable binpatch minimization (#2302)

Co-authored-by: Sebastian Hartte <shartte@users.noreply.github.com>

Technici4n committed:
Reduce collisions in PatchedDataComponentMap hashCode (#2332)
